### ISSUE TITLE
grub-btrfsd started, waiting for snapshots but does not detect any

### ISSUE THREAD

**Issue by FGD-Garuda** [Antynea/grub-btrfs#261]
Hello there,

I did all I could to avoid posting but here I am.

I come from the old grub-btrfs way (when using `grub-btrfs.path`) and been using it for 4 years I think.
I am on Arch, NOT using timeshift/snapper and all my snapshots are in the default `/.snapshots` path.
I updated to `core/btrfs-progs 6.1.2` after seeing issues here related to that.
~~~
╰─λ sudo /etc/grub.d/41_snapshots-btrfs --version
Version 4.12
~~~

~~~
─λ sudo cat /etc/default/grub-btrfs/config
#!/usr/bin/env bash

GRUB_BTRFS_VERSION=4.12

# Disable grub-btrfs.
# Default: "false"
#GRUB_BTRFS_DISABLE="true"

# Name appearing in the Grub menu.
# Default: "Use distribution information from /etc/os-release."
GRUB_BTRFS_SUBMENUNAME="Snapshots"

# Custom title.
# Shows/Hides "date" "snapshot" "type" "description" in the Grub menu, custom order available.
# Default: ("date" "snapshot" "type" "description")
#GRUB_BTRFS_TITLE_FORMAT=("date" "snapshot" "type" "description")

# Limit the number of snapshots populated in the GRUB menu.
# Default: "50"
#GRUB_BTRFS_LIMIT="50"

# Sort the found subvolumes by "ogeneration" or "generation" or "path" or "rootid".
# # See Sorting section to https://btrfs.wiki.kernel.org/index.php/Manpage/btrfs-subvolume#SUBCOMMAND
# "-rootid" means list snapshot by new ones first.
# Default: "-rootid"
#GRUB_BTRFS_SUBVOLUME_SORT="+ogen,-gen,path,rootid"

# Show snapshots found during run "grub-mkconfig"
# Default: "true"
#GRUB_BTRFS_SHOW_SNAPSHOTS_FOUND="false"

# Show Total of snapshots found during run "grub-mkconfig"
# Default: "true"
#GRUB_BTRFS_SHOW_TOTAL_SNAPSHOTS_FOUND="true"

# By default, "grub-btrfs" automatically detects most existing kernels.
# If you have one or more custom kernels, you can add them here.
# Default: ("")
#GRUB_BTRFS_NKERNEL=("kernel-custom" "vmlinux-custom")

# By default, "grub-btrfs" automatically detects most existing initramfs.
# If you have one or more custom initramfs, you can add them here.
# Default: ("")
#GRUB_BTRFS_NINIT=("initramfs-custom.img" "initrd-custom.img" "otherinit-custom.gz")

# By default, "grub-btrfs" automatically detects most existing microcodes.
# If you have one or more custom microcodes, you can add them here.
# Default: ("")
#GRUB_BTRFS_CUSTOM_MICROCODE=("custom-ucode.img" "custom-uc.img "custom_ucode.cpio")

# Comma seperated mount options to be used when booting a snapshot.
# They can be defined here as well as in the "/" line inside the respective snapshots'
# "/etc/fstab" files.  Mount options found in both places are combined, and this variable
# takes priority over `fstab` entries.
# NB: Do NOT include "subvol=..." or "subvolid=..." here.
# Default: ""
#GRUB_BTRFS_ROOTFLAGS="space_cache,commit=10,norecovery"

# Ignore specific path during run "grub-mkconfig".
# Only exact paths are ignored.
# e.g : if `specific path` = @, only `@` snapshot will be ignored.
# Default: ("@")
GRUB_BTRFS_IGNORE_SPECIFIC_PATH=("@")

# Ignore prefix path during run "grub-mkconfig".
# Any path starting with the specified string will be ignored.
# e.g : if `prefix path` = @, all snapshots beginning with "@/..." will be ignored.
# Default: ("var/lib/docker" "@var/lib/docker" "@/var/lib/docker")
GRUB_BTRFS_IGNORE_PREFIX_PATH=("var/lib/docker" "@var/lib/docker" "@/var/lib/docker")

# Ignore specific type/tag of snapshot during run "grub-mkconfig".
# For snapper:
# Type = single, pre, post.
# For Timeshift:
# Tag = boot, ondemand, hourly, daily, weekly, monthly.
# Default: ("")
#GRUB_BTRFS_IGNORE_SNAPSHOT_TYPE=("")

# Ignore specific description of snapshot during run "grub-mkconfig".
# e.g: timeline
# Default: ("")
#GRUB_BTRFS_IGNORE_SNAPSHOT_DESCRIPTION=("")

# By default "grub-btrfs" automatically detects your boot partition,
# either located at the system root or on a separate partition or in a subvolume,
# Change to "true" if your boot partition isn't detected as separate.
# Default: "false"
#GRUB_BTRFS_OVERRIDE_BOOT_PARTITION_DETECTION="true"

# Location of the folder containing the "grub.cfg" file.
# Use by grub-btrfs to save the file "grub-btrfs.cfg".
# Might be grub2 on some systems.
# For example, on Fedora with EFI : "/boot/efi/EFI/fedora"
# Default: "/boot/grub"
#GRUB_BTRFS_GRUB_DIRNAME="/boot/grub2"

# Location of kernels/initramfs/microcode.
# Use by "grub-btrfs" to detect the boot partition and the location of kernels/initrafms/microcodes.
# Default: "/boot"
#GRUB_BTRFS_BOOT_DIRNAME="/boot"

# Name/path of grub-mkconfig command, use by "grub-btrfs.service"
# Might be 'grub2-mkconfig' on some systems (Fedora ...)
# Default paths are /sbin:/bin:/usr/sbin:/usr/bin,
# if your path is missing, report it on the upstream project.
# For example, on Fedora : "/sbin/grub2-mkconfig"
# You can use only name or full path.
# Default: grub-mkconfig
#GRUB_BTRFS_MKCONFIG=/usr/bin/grub2-mkconfig

# Name of grub-script-check command, use by "grub-btrfs"
# Might be 'grub2-script-check' on some systems (Fedora ...)
# For example, on Fedora : "grub2-script-check"
# Default: grub-script-check
#GRUB_BTRFS_SCRIPT_CHECK=grub2-script-check

# Path of grub-mkconfig_lib file, use by "grub-btrfs"
# Might be '/usr/share/grub2/grub-mkconfig_lib' on some systems (Opensuse ...)
# Default: /usr/share/grub/grub-mkconfig_lib
#GRUB_BTRFS_MKCONFIG_LIB=/usr/share/grub2/grub-mkconfig_lib

# Password protection management for submenu,snapshots
# Refer to the Grub documentation https://www.gnu.org/software/grub/manual/grub/grub.html#Authentication-and-authorisation
# and this comment https://github.com/Antynea/grub-btrfs/issues/95#issuecomment-682295660
#
# Add authorized usernames separate by comma (foo,bar)
# When Grub's password protection is enabled, the superuser is authorized by default, it isn't necessary to add it
# Default: ""
#GRUB_BTRFS_PROTECTION_AUTHORIZED_USERS="foo,bar"
#
# Disable authentication support for submenu of Grub-btrfs only (--unrestricted)
# doesn't work if GRUB_BTRFS_PROTECTION_AUTHORIZED_USERS isn't empty
# Default: "false"
#GRUB_BTRFS_DISABLE_PROTECTION_SUBMENU="true"
~~~

I notice I still have `/etc/systemd/system/\x2esnapshots.mount.wants/grub-btrfs.path` on the system, which does not exist when I try to delete it. I see it with the `find` command.

My issue is:
When the computer starts, I get this:

~~~
╰─λ sudo systemctl status grub-btrfsd
● grub-btrfsd.service - Regenerate grub-btrfs.cfg
Loaded: loaded (/etc/systemd/system/grub-btrfsd.service; enabled; preset: disabled)
Active: active (running) since Sat 2023-01-07 10:53:09 EST; 27s ago
Main PID: 626 (grub-btrfsd)
Tasks: 2 (limit: 18452)
Memory: 1016.0K
CPU: 9ms
CGroup: /system.slice/grub-btrfsd.service
├─626 /bin/sh /usr/bin/grub-btrfsd --syslog /.snapshots
└─637 inotifywait -q -q -e create -e delete -e unmount -t 0 /.snapshots

jan 07 10:53:09 rene-hm90 grub-btrfsd[632]: grub-btrfsd starting up...
~~~

Then I either delete snapshots or create new ones and nothing changes, `/boot/grub/grub-btrfs.cfg` does not get regenerated.

I tried running with verbose and I get:

~~~
╰─λ sudo /usr/bin/grub-btrfsd /.snapshots --verbose
grub-btrfsd starting up...
Watching /.snapshots for new snapshots...
~~~

Again no further msg when I create new snapshots.
If I restart the daemon `sudo systemctl restart grub-btrfsd`, I get 1 step closer, but still not working:

~~~
╰─λ sudo systemctl status grub-btrfsd
● grub-btrfsd.service - Regenerate grub-btrfs.cfg
Loaded: loaded (/etc/systemd/system/grub-btrfsd.service; enabled; preset: disabled)
Active: active (running) since Sat 2023-01-07 11:25:11 EST; 1s ago
Main PID: 218510 (grub-btrfsd)
Tasks: 2 (limit: 18452)
Memory: 848.0K
CPU: 14ms
CGroup: /system.slice/grub-btrfsd.service
├─218510 /bin/sh /usr/bin/grub-btrfsd --syslog /.snapshots
└─218517 inotifywait -q -q -e create -e delete -e unmount -t 0 /.snapshots

jan 07 11:25:11 rene-hm90 grub-btrfsd[218514]: grub-btrfsd starting up...
jan 07 11:25:11 rene-hm90 grub-btrfsd[218516]: Watching /.snapshots for new snapshots...
~~~

If I stop, disable, enable and start the daemon, I get to the same point as a reboot:

~~~
╰─λ sudo systemctl stop grub-btrfsd

╰─λ sudo systemctl disable grub-btrfsd
Removed "/etc/systemd/system/multi-user.target.wants/grub-btrfsd.service".

╰─λ sudo systemctl enable grub-btrfsd
Created symlink /etc/systemd/system/multi-user.target.wants/grub-btrfsd.service → /etc/systemd/system/grub-btrfsd.service.

╰─λ sudo systemctl start grub-btrfsd

╰─λ sudo systemctl status grub-btrfsd
● grub-btrfsd.service - Regenerate grub-btrfs.cfg
Loaded: loaded (/etc/systemd/system/grub-btrfsd.service; enabled; preset: disabled)
Active: active (running) since Sat 2023-01-07 11:28:22 EST; 36s ago
Main PID: 239629 (grub-btrfsd)
Tasks: 2 (limit: 18452)
Memory: 840.0K
CPU: 16ms
CGroup: /system.slice/grub-btrfsd.service
├─239629 /bin/sh /usr/bin/grub-btrfsd --syslog /.snapshots
└─239640 inotifywait -q -q -e create -e delete -e unmount -t 0 /.snapshots

jan 07 11:28:22 rene-hm90 grub-btrfsd[239637]: grub-btrfsd starting up...
~~~

I have not edited `grub-btrfsd.service` at all and it points to `/.snapshots`.
Curiously I was able once to make it work while playing around with it today and I saw snapshots detection msgs in the verbose, but then after a reboot to see if it'll hold up, I can no longer make it detect snapshots.

Not sure what to do from here...

Thanks!

**Comment by FGD-Garuda** [comment]
I am adding some more information.
I still got remnants of `grub-btrfs` old way of working. 

~~~
╰─λ ls /etc/systemd/system/
drwxr-xr-x   - root 12 mar  2022  \x2esnapshots.mount.wants
drwxr-xr-x   - root 12 mar  2022  grub-btrfs.path.d

╰─λ ls /etc/systemd/system/grub-btrfs.path.d/
.rw-r--r-- 419 root 17 mar  2022  override.conf

╰─λ  ls "/etc/systemd/system/\x2esnapshots.mount.wants/"
lrwxrwxrwx 39 root 12 mar  2022  grub-btrfs.path -> /usr/lib/systemd/system/grub-btrfs.path
~~~
The target of the symlink does not exist anymore.

If I start the old `grub-btrfs`, it works and generates new snapshots. But it cannot be enabled on boot cuz the unit has no installation config, which is what systemctl tells me. I guess I could create a service to start that service, but that's not how it should be done. loll Anyway the old grub-btrfs.service is not maintained anymore AFAIK.

Maybe I don't have a clean installation of `grub-btrfsd`?

**Comment by FGD-Garuda** [comment]
HA!! Found the issue... the `/.snapshots` target folder is non-recursive, just like the original grub-btrfs.

90% of my snapshots are in SUB-folders of /.snapshots, which is why it worked only once, when I created a snapshot directly in /.snapshots.

Ok so now how do I add multiple target folders to watch?
Multiple `ExecStart` lines?
I tried just adding the folders at the end of the line, just like this `ExecStart=/usr/bin/grub-btrfsd --syslog /.snapshots /.snapshots/ad-hoc`, but it's not catching the /ad-hoc sub-folder.
I'll read the man see if I find anything.
